docs: Retry documentation (#194)
confluentinc/parallel-consumer: README.adoc

Even during retries, offsets will always be committed only after successful processing, and in order.

== Retries

> If processing of a record fails, the record will be placed back into it's queue and retried after a time which can be configured (see the `ParallelConsumerOptions` class).

Reviewer: ```suggestion
If processing of a record fails, the record will be placed back into it's queue and retried with a configurable delay (see the `ParallelConsumerOptions` class).
```

Ordering guarantees will always be adhered to, regardless of failure.

A failure is denoted by *any* exception being thrown from the user's processing function.
The system catches these exceptions, logs them and replaces the record in the queue for processing later.
All types of Exceptions thrown are considered retriable.
To not retry a record, do not throw an exception from your processing fuction.

If for some reason you want to proactively fail a record, without relying on some other system throwing an exception which you don't catch - simply throw an exception of your own design, which the system will treat the same way.

To configure the retry delay, see `ParallelConsumerOptions#defaultRetryDelay`.

At the moment there is no terminal error support, so messages will continue to be retried forever as long as an exception continues to be thrown from the user function (see <<skipping-records>>).

> But still this will not hold up the queues in `KEY` or `UNORDERED` modes, however `PARTITION` mode it *will* block progress.

Reviewer: ```suggestion
But still this will not hold up the queues in `KEY` or `UNORDERED` modes, however in `PARTITION` mode it *will* block progress.
```

If for whatever reason you want to skip a record, simply do not throw an exception, or catch any exception being thrown, log and swallow it and return from the user function normally.
The system will treat this as a record processing success, mark the record as completed and move on as though it were normal operation.

A user may choose to skip a record for example, if it has been retried too many times or if the record is invalid or doesn't need processing.

Implementing a https://github.com/confluentinc/parallel-consumer/issues/196[max retries feature] as a part of the system is planned.

.Example of skipping a record after a maximum number of retries is reached
[source,java,indent=0]
----
        final int maxRetries = 10;
        final Map<ConsumerRecord<String, String>, Long> retriesCount = new ConcurrentHashMap<>();

        pc.poll(consumerRecord -> {
            Long retryCount = retriesCount.computeIfAbsent(consumerRecord, ignore -> 0L);
            if (retryCount < maxRetries) {
                processRecord(consumerRecord);
                // no exception, so completed - remove from map
                retriesCount.remove(consumerRecord);
            } else {

>                 log.warn("Retry count {} exceed manx of {} for record {}", retryCount, maxRetries, consumerRecord);

Reviewer: ```suggestion
                log.warn("Retry count {} exceeded max of {} for record {}", retryCount, maxRetries, consumerRecord);
```

                // giving up, remove from map
                retriesCount.remove(consumerRecord);
            }
        });
----

=== Circuit Breaker Pattern

> Althought the system doesn't have an https://github.com/confluentinc/parallel-consumer/issues/110[explicit circuit breaker pattern feature], one can be created by conbining the custom retry delay function and proactive failure.

Reviewer: ```suggestion
Although the system doesn't have an https://github.com/confluentinc/parallel-consumer/issues/110[explicit circuit breaker pattern feature], one can be created by combining the custom retry delay function and proactive failure.
```
